### PR TITLE
Completely overhaul list display

### LINKED ISSUE #536: Don't deselect songs after choosing an action
When I select a number of songs and then pin them, 
the selection is cancelled. If I also wanted to add them to a playlist, I have to reselect them.

Todo:
* Also refactor TrackView to use the GenericModels

### PR DESCRIPTION
**Pull request by tzugen**
Changes so far:

- Remove UpdateView code
- Add more undestandable icons for Add Next/Add Last actions
- Bookmarks section gets a header view like other Track Collections
- Show a toast when deleting/Unpinning songs
- Keep selection after choosing a button (fixes #536)
- If no songs are selected, pressing delete doesn't delete all visible songs anymore :scream: 
- Refactor the folder selector into its own distinct component
- Improve the display of the download status
- Fix jumping songs in Download view
- Remove unmaintained sortListView
- Add pretty views if lists are empty

Screenshot_20211202-123543_Ultrasonic

Screenshot_20211202-123503_Ultrasonic

### CODE REVIEW

**Comment by tzugen** [comment]
I think I have it done.

@nitehu I tried to avoid touching any code which handles files, so that you can merge easily, but I had to touch OfflineMusicService in a few places, hope it won't be too much work when you merge the file code...

**Comment by tzugen** [comment]
And should I flatten into a single commit? What do you think?

**Comment by nitehu** [comment]
I had no time to progress lately, but I'll try to check this on the weekend.
As for the squash, I really have no preference... As I see in the history there were some PRs with even more commits before. Sometimes the separate commit messages can be useful when later someone tries to find out why was something changed.

**Comment by nitehu** [comment]
OK I've checked the modifications in the files, and now I'm pretty sure you shouldn't squash :) In an ideal world these changes would go into at least 4-6 separate PRs (of course I know in reality it isn't that easy to change only one thing in Ultrasonic at once... so I think I understand why this is one big PR)

**Comment by nitehu** [comment]
Something isn't right with the layout of the playlist on the Now Playing screen.
I'm not sure how this can be reproduced, anyways here is a shot:
Screenshot 2021-12-05 170015
The track number's padding sometimes gets this large. Also, somehow I got the "100%" stuck, it was still displayed after the song was downloaded and started playing.  


Screenshot 2021-12-05 170758
The vertical padding (/size) of the items is also off, there are smaller and bigger items in the list, or at least they aren't centered correctly.  

Screenshot 2021-12-05 171418
If the length of the track gets double digits, it pushes the arrow to the left

**Comment by nitehu** [comment]
Screenshot 2021-12-05 171947

The drag handle doesn't work for me, but when I try to drag the item e.g. in its center, I can drag it, but also the context menu appears :D

Edit: It seems this also isn't the correct context menu and it isn't even working...
Also, during dragging it is more visible that an item shouldn't have transparent background.

**Comment by nitehu** [comment]
- Selecting an album in Ultrasonic Main -> Albums, and then pressing Back will take to a seemingly random place in the Albums list


Screenshot 2021-12-05 174659
- Selecting an item in Ultrasonic Main -> Genres just shows an empty screen. Sometimes at the bottom there is an arrow, which takes to another empty page.

- The Refresh at Ultrasonic Main -> Random Songs works OK, but Refreshing the Ultrasonic Main -> Random Albums updates the list by moving the items around, which is a bit confusing

**Comment by nitehu** [comment]
I've seen a fail loop during playing some albums / songs, but I can't reliably reproduce it.
I'll copy the logs here, maybe you've seen it before or will have better luck finding what's this:
[log.txt](https://github.com/ultrasonic/ultrasonic/files/7656448/log.txt)

Eidt: OK the situation got better and worse at the same time, now the emulator always stucks in this loop. I can't continue testing this way, but maybe tomorrow I can find out what caused this exactly...

**Comment by tzugen** [comment]
> * Selecting an album in Ultrasonic Main -> Albums, and then pressing Back will take to a seemingly random place in the Albums list
> 
> 
> [Screenshot 2021-12-05 174659](https://user-images.githubusercontent.com/55684536/144755466-4be1ac38-b686-4e6c-a0af-6c294d468df7.jpg)
> 
>     * Selecting an item in Ultrasonic Main -> Genres just shows an empty screen. Sometimes at the bottom there is an arrow, which takes to another empty page.
> 
>     * The Refresh at Ultrasonic Main -> Random Songs works OK, but Refreshing the Ultrasonic Main -> Random Albums updates the list by moving the items around, which is a bit confusing

These three should be fixed now.

**Comment by tzugen** [comment]
> [Screenshot 2021-12-05 171947](https://user-images.githubusercontent.com/55684536/144754539-300e6adc-ad09-477e-8d31-6636458ca88a.jpg)
> 
> The drag handle doesn't work for me, but when I try to drag the item e.g. in its center, I can drag it, but also the context menu appears :D
> 
> Edit: It seems this also isn't the correct context menu and it isn't even working... Also, during dragging it is more visible that an item shouldn't have transparent background.

The context menu shouldn't have been shown, it's fixed now. Regarding the drag handle, it seems that this seems to be the default with this ``ItemTouchHelper`` that you drag the whole item. I left the drag handles as a visual indication that you can drag the item.

I also gave the item that is being moved 0.6f alpha transparency to avoid the overlap.

**Comment by tzugen** [comment]
> Something isn't right with the layout of the playlist on the Now Playing screen. I'm not sure how this can be reproduced, anyways here is a shot: [Screenshot 2021-12-05 170015](https://user-images.githubusercontent.com/55684536/144754191-689c7873-1c18-4b67-a300-a1e4eb973377.jpg) The track number's padding sometimes gets this large. Also, somehow I got the "100%" stuck, it was still displayed after the song was downloaded and started playing.
> 
> [Screenshot 2021-12-05 170758](https://user-images.githubusercontent.com/55684536/144754263-9a5c3b49-77dc-4f06-9373-a60dd5f53d78.jpg) The vertical padding (/size) of the items is also off, there are smaller and bigger items in the list, or at least they aren't centered correctly.
> 
> [Screenshot 2021-12-05 171418](https://user-images.githubusercontent.com/55684536/144754349-4dd9b3e6-a24a-4145-af6b-747d90225551.jpg) If the length of the track gets double digits, it pushes the arrow to the left

The last one I could fix. The two before I couldn't reproduce...

File: ultrasonic/src/main/kotlin/org/moire/ultrasonic/adapters/TrackViewHolder.kt [hunk 1]
```diff
@@ -112,11 +115,15 @@ class TrackViewHolder(val view: View) : RecyclerView.ViewHolder(view), Checkable
             progress.isVisible = false
         }
 
-        RxBus.playerStateObservable.subscribe {
+        rxSubscription = RxBus.playerStateObservable.subscribe {
```
Comment (nitehu) [inline ultrasonic/src/main/kotlin/org/moire/ultrasonic/adapters/TrackViewHolder.kt:118]: I'm not sure how this worked without unsubscribing, but doing it explicitly feels safer

File: ultrasonic/src/main/kotlin/org/moire/ultrasonic/adapters/TrackViewBinder.kt [hunk 2]
```diff
@@ -89,6 +96,13 @@ class TrackViewBinder(
             }
         }
 
+        holder.drag.setOnTouchListener { _, event ->
```
Comment (nitehu) [inline ultrasonic/src/main/kotlin/org/moire/ultrasonic/adapters/TrackViewBinder.kt:97]: ItemTouchHelper can use a drag handle, and in some samples they do it like this.

File: ultrasonic/src/main/res/layout/list_item_track_details.xml [hunk 3]
```diff
@@ -62,7 +64,7 @@
         app:layout_constraintBottom_toBottomOf="parent"
         app:layout_constraintEnd_toStartOf="@+id/song_duration"
         app:layout_constraintStart_toStartOf="parent"
-        app:layout_constraintTop_toBottomOf="@+id/song_track"
+        app:layout_constraintTop_toBottomOf="@+id/song_title"
```
Comment (nitehu) [inline ultrasonic/src/main/res/layout/list_item_track_details.xml]: This caused some layout problems when the track number was missing

**Comment by nitehu** [comment]
Fortunately the error loop with the media player may just have been caused by the emulator, I've cold booted it and that seems to fixed that.

I'll test this PR more and see if I can find anything else.

**Comment by tzugen** [comment]
> Fortunately the error loop with the media player may just have been caused by the emulator, I've cold booted it and that seems to fixed that.
> 
> I'll test this PR more and see if I can find anything else.

Yes I thought the same, the log you gave me showed an Error in the actual platform MediaPlayer amd I don't think I touched any code there.

Nice work on the drag handler! I increased its paddings a bit to give it a larger area for dragging.

**Comment by nitehu** [comment]
Oh! I've just noticed, the playlist is missing its scrollbar...
Also, the artist list had a fastscroll thumb which displayed the current letter during scrolling, and now that's gone too. Not that I miss it much, I'm just not sure if it's intentionally gone.

Everything else seems to work now correctly.

**Comment by tzugen** [comment]
> Oh! I've just noticed, the playlist is missing its scrollbar... Also, the artist list had a fastscroll thumb which displayed the current letter during scrolling, and now that's gone too. Not that I miss it much, I'm just not sure if it's intentionally gone.
> 
> Everything else seems to work now correctly.

Both fixed :)

**Review by nitehu — APPROVED** [review]
I think now everything works well, and we can always open a new issue if we find any problems.
Code review was also OK, so we can merge this.

**Comment by tzugen** [comment]
# :partying_face: 

thanks for taking the time to review!

**Comment by nitehu** [comment]
👍 and thanks for your hard work!

**Pull request merged**